src/MongoDB/Manager.c

```diff
@@ -200,6 +200,15 @@ static void php_phongo_manager_prep_uri_options(zval *options TSRMLS_DC) /* {{{
 				continue;
 			}
 
+            if (!strcasecmp(ZSTR_VAL(string_key), MONGOC_URI_READPREFERENCE)) {
+				ZVAL_DEREF(option);
+				SEPARATE_ZVAL_NOREF(option);
+				if (Z_TYPE_P(option) != IS_STRING) {
+                    phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Unsupported readPreference type, a string must be given");
```

Comment:
> as we will implement a new validation function that applied on BSON values instead of zval

I'm suggesting that you not modify anything in `Manager.c` and instead focus on `php_phongo_apply_options_to_uri()` (and similar function). Consider:

```c
if (mongoc_uri_option_is_int32(key)) {
	if (!BSON_ITER_HOLDS_INT32(&iter)) {
		/* phongo_throw_exception() for invalid type */
		return false;
	}

	if (!mongoc_uri_set_option_as_int32(uri, key, bson_iter_int32(&iter))) {
		phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Failed to parse \"%s\" URI option", key);
		return false;
	}

	continue;
}
```

> another question about validating bson value's types

There are macros in libbson that check types. `BSON_ITER_HOLDS_INT32()`, which appears in the example above, is one of these macros. When we're working with BSON, libbson provides a `bson_iter_t` that can point to a specific field within a document/array. The common API use case is to move the iter to the field corresponding to a given name, and then we can check its type code or access its value. In cases where the value is another BSON document or array, we can also recurse the iter a level deeper (not relevant to this ticket but I wanted to mention it).